lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -299,10 +307,24 @@ def create_opensearch_users(
             index: the index to which the users must be granted access
             extra_user_roles: the level of permissions that the user should be given. Can be a
                 comma-separated list of roles, which should result in a merged list of permissions.
+            relation_id: the relation id for this relation, if it exists
 
         Raises:
             OpenSearchUserMgmtError if user creation fails
         """
+        rel_id = str(relation_id) if relation_id is not None else None
+
+        users = json.loads(
+            self.charm.peers_data.get(Scope.APP, ClientUsersDict, auto_casting=False) or "{}"
+        )
+        if rel_id not in users:
+            users[rel_id] = []
+        roles = json.loads(
+            self.charm.peers_data.get(Scope.APP, ClientRolesDict, auto_casting=False) or "{}"
+        )
+        if rel_id not in roles:
+            roles[rel_id] = []
```

Comment:
1. Why do we store the `roles` - given the role is exactly the username? 
2. If really necessary - Could we merge these 2 into a single dict: `{"rel_id": {"roles": ..., "users": ....}}`

Reply:
Solution fully re-worked: closing this thread.